ntoskrnl/include/internal/i386/ke.h

```diff
@@ -722,7 +722,7 @@ KiSystemCallTrampoline(IN PVOID Handler,
         : "c"(StackBytes),
           "d"(Arguments),
           "r"(Handler)
-        : "%esp", "%esi", "%edi"
+        : "%esi", "%edi"
```

Comment:
Please leave the GCC compiler code untouched as is. For specific compiler system call handling code you should do that separately, in `__clang__` respectively.

Reply:
Can you give a disassembly of clang output  before & after your change here ?

Reply:
> 
> 
> Please leave the GCC compiler code untouched as is. For specific compiler system call handling code you should do that separately, in `__clang__` respectively.

OK.



> 
> 
> Can you give a disassembly of clang output before & after your change here ?

Yes, I will. I've added a change I had to make before the one you commented. I've commited what was on my mind at first.

Reply:
I've added a change I had to make before the one you commented. I've commited what was on my mind at first.

Reply:
> 
> 
> You need to look up predefined macros for both clang and gcc on command line, like:
> 
> ```
> $ echo | clang -dM -E -
> ```
> 
> and
> 
> ```
> $ echo | gcc -dM -E -
> ```
> 
> I can fix this up with opening pr on to your working branch for help.



> 
> 
> Give an example with opening a PR onto valeriy-zainullin:clang-build-boots branch. Use this type of approachs.

I think that I did what @GeoB99 has asked. Should I still do what you suggest?

Reply:
Here's disassemblies before the change and after.

<details>
  <summary>Disassembly before the change</summary>

```
.text:004F6BD0 ; =============== S U B R O U T I N E =======================================
.text:004F6BD0
.text:004F6BD0 ; Attributes: bp-based frame fuzzy-sp
.text:004F6BD0
.text:004F6BD0 sub_4F6BD0      proc near               ; CODE XREF: sub_403D39+62↑p
.text:004F6BD0                                         ; sub_403D39+F6↑p ...
.text:004F6BD0                 push    ebp
.text:004F6BD1                 mov     ebp, esp
.text:004F6BD3                 push    ebx
.text:004F6BD4                 push    edi
.text:004F6BD5                 push    esi
.text:004F6BD6                 and     esp, 0FFFFFFFCh
.text:004F6BD9                 sub     esp, 18h
.text:004F6BDC                 mov     esi, esp
.text:004F6BDE                 mov     [esi+14h], edx
.text:004F6BE1                 mov     ebx, ecx
.text:004F6BE3                 mov     eax, [ecx+3Ch]
.text:004F6BE6                 mov     ecx, [ecx+44h]
.text:004F6BE9                 mov     edx, ds:0FFDFF124h
.text:004F6BEF                 mov     [ebx+0Ch], eax
.text:004F6BF2                 mov     dword ptr [ebx+8], 0BADB0D00h
.text:004F6BF9                 mov     eax, [ebx+68h]
.text:004F6BFC                 mov     [ebx+4], eax
.text:004F6BFF                 mov     eax, [ebx+60h]
.text:004F6C02                 mov     [ebx], eax
.text:004F6C04                 mov     dword ptr [ebx+48h], 0FFFFFFFFh
.text:004F6C0B                 mov     eax, [edx+110h]
.text:004F6C11                 mov     [ebx+3Ch], eax
.text:004F6C14                 mov     dword ptr [ebx+64h], 0
.text:004F6C1B                 movsx   eax, byte ptr [edx+0D7h]
.text:004F6C22                 mov     [ebx+48h], eax
.text:004F6C25                 mov     eax, ds:0FFDFF000h
.text:004F6C2A                 mov     [ebx+4Ch], eax
.text:004F6C2D                 mov     dword ptr ds:0FFDFF000h, 0FFFFFFFFh
.text:004F6C37                 mov     dword ptr [ebx+2Ch], 0
.text:004F6C3E                 test    byte ptr [ebx+6Ch], 1
.text:004F6C42                 jz      short loc_4F6CB8
.text:004F6C44                 mov     eax, ds:0FFDFF124h
.text:004F6C49                 cmp     byte ptr [eax+3], 0
.text:004F6C4D                 jz      short loc_4F6CB8
.text:004F6C4F                 mov     eax, large fs:20h
.text:004F6C55                 mov     edi, ecx
.text:004F6C57                 mov     ecx, dr0
.text:004F6C5A                 mov     [ebx+18h], ecx
.text:004F6C5D                 mov     ecx, dr1
.text:004F6C60                 mov     [ebx+1Ch], ecx
.text:004F6C63                 mov     ecx, dr2
.text:004F6C66                 mov     [ebx+20h], ecx
.text:004F6C69                 mov     ecx, dr3
.text:004F6C6C                 mov     [ebx+24h], ecx
.text:004F6C6F                 mov     ecx, dr6
.text:004F6C72                 mov     [ebx+28h], ecx
.text:004F6C75                 mov     ecx, dr7
.text:004F6C78                 mov     [ebx+2Ch], ecx
.text:004F6C7B                 xor     ecx, ecx
.text:004F6C7D                 mov     dr7, ecx
.text:004F6C80                 mov     ecx, [eax+2F8h]
.text:004F6C86                 mov     dr0, ecx
.text:004F6C89                 mov     ecx, [eax+2FCh]
.text:004F6C8F                 mov     dr1, ecx
.text:004F6C92                 mov     ecx, [eax+300h]
.text:004F6C98                 mov     dr2, ecx
.text:004F6C9B                 mov     ecx, [eax+304h]
.text:004F6CA1                 mov     dr3, ecx
.text:004F6CA4                 mov     ecx, [eax+308h]
.text:004F6CAA                 mov     dr6, ecx
.text:004F6CAD                 mov     ecx, edi
.text:004F6CAF                 mov     eax, [eax+30Ch]
.text:004F6CB5                 mov     dr7, eax
.text:004F6CB8
.text:004F6CB8 loc_4F6CB8:                             ; CODE XREF: sub_4F6BD0+72↑j
.text:004F6CB8                                         ; sub_4F6BD0+7D↑j
.text:004F6CB8                 mov     [edx+110h], ebx
.text:004F6CBE                 mov     al, [ebx+6Ch]
.text:004F6CC1                 and     al, 1
.text:004F6CC3                 mov     [edx+0D7h], al
.text:004F6CC9                 sti
.text:004F6CCA                 mov     edi, ecx
.text:004F6CCC                 shr     edi, 8
.text:004F6CCF                 and     edi, 10h
.text:004F6CD2                 mov     [esi+0Ch], ecx
.text:004F6CD5                 and     ecx, 0FFFh
.text:004F6CDB                 mov     [esi+4], edx
.text:004F6CDE                 mov     edx, [edx+118h]
.text:004F6CE4                 mov     eax, ecx
.text:004F6CE6                 mov     [esi+8], ecx
.text:004F6CE9                 cmp     ecx, [edx+edi+8]
.text:004F6CED                 jnb     loc_4F6D7D
.text:004F6CF3                 add     edx, edi
.text:004F6CF5                 mov     [esi+10h], edx
.text:004F6CF8
.text:004F6CF8 loc_4F6CF8:                             ; CODE XREF: sub_4F6BD0+27B↓j
.text:004F6CF8                 test    edi, edi
.text:004F6CFA                 jz      short loc_4F6D12
.text:004F6CFC                 mov     eax, large fs:18h
.text:004F6D02                 cmp     dword ptr [eax+0F70h], 0
.text:004F6D09                 jz      short loc_4F6D12
.text:004F6D0B                 mov     eax, ds:off_58D848
.text:004F6D10                 call    dword ptr [eax]
.text:004F6D12
.text:004F6D12 loc_4F6D12:                             ; CODE XREF: sub_4F6BD0+12A↑j
.text:004F6D12                                         ; sub_4F6BD0+139↑j
.text:004F6D12                 mov     eax, large fs:20h
.text:004F6D18                 inc     dword ptr [eax+598h]
.text:004F6D1E                 mov     eax, ds:off_58DBC4
.text:004F6D23                 mov     edi, [esi+14h]
.text:004F6D26                 cmp     [eax], edi
.text:004F6D28                 mov     eax, [esi+10h]
.text:004F6D2B                 mov     eax, [eax+0Ch]
.text:004F6D2E                 mov     ecx, [esi+8]
.text:004F6D31                 movzx   eax, byte ptr [eax+ecx]
.text:004F6D35                 jbe     short loc_4F6D93
.text:004F6D37                 test    byte ptr [ebx+6Ch], 1
.text:004F6D3B                 jnz     short loc_4F6D93
.text:004F6D3D                 push    72Ch
.text:004F6D42                 push    offset aNtoskrnlKeI386_7 ; "ntoskrnl/ke/i386/traphdlr.c"
.text:004F6D47                 push    offset aKisystemservic ; "KiSystemServiceHandler"
.text:004F6D4C                 push    offset aUnrecoverableE_0 ; "UNRECOVERABLE ERROR:  %s at %s:%d is UN"...
.text:004F6D51                 call    DbgPrint
.text:004F6D56                 add     esp, 10h
.text:004F6D59                 push    offset unk_56DC42
.text:004F6D5E                 call    DbgPrint
.text:004F6D63                 add     esp, 4
.text:004F6D66                 call    DbgBreakPoint
.text:004F6D6B                 push    0
.text:004F6D6D                 push    0
.text:004F6D6F                 push    0
.text:004F6D71                 push    0C0000002h
.text:004F6D76                 push    5Ah
.text:004F6D78                 call    KeBugCheckEx
.text:004F6D7D
.text:004F6D7D loc_4F6D7D:                             ; CODE XREF: sub_4F6BD0+11D↑j
.text:004F6D7D                 test    edi, edi
.text:004F6D7F                 jnz     loc_4F6E22
.text:004F6D85                 mov     dword ptr [esi], 0C000001Ch
.text:004F6D8B                 mov     edx, [esi+4]
.text:004F6D8E                 jmp     loc_4F6E55
.text:004F6D93 ; ---------------------------------------------------------------------------
.text:004F6D93
.text:004F6D93 loc_4F6D93:                             ; CODE XREF: sub_4F6BD0+165↑j
.text:004F6D93                                         ; sub_4F6BD0+16B↑j
.text:004F6D93                 mov     [esi], eax
.text:004F6D95                 cmp     dword ptr [esi+0Ch], 1000h
.text:004F6D9C                 jb      short loc_4F6DAD
.text:004F6D9E                 mov     eax, ds:dword_597008
.text:004F6DA3                 test    eax, eax
.text:004F6DA5                 jz      short loc_4F6DAD
.text:004F6DA7                 push    edi
.text:004F6DA8                 push    dword ptr [esi+0Ch]
.text:004F6DAB                 call    eax ; dword_597008
.text:004F6DAD
.text:004F6DAD loc_4F6DAD:                             ; CODE XREF: sub_4F6BD0+1CC↑j
.text:004F6DAD                                         ; sub_4F6BD0+1D5↑j
.text:004F6DAD                 mov     eax, [esi+10h]
.text:004F6DB0                 mov     eax, [eax]
.text:004F6DB2                 mov     ecx, [esi+8]
.text:004F6DB5                 mov     eax, [eax+ecx*4]
.text:004F6DB8                 mov     ecx, [esi]
.text:004F6DBA                 mov     edx, edi
.text:004F6DBC                 sub     esp, ecx        ; The start of "KiSystemCallTrampoline".
.text:004F6DBE                 mov     edi, esp
.text:004F6DC0                 mov     esi, edx
.text:004F6DC2                 shr     ecx, 2
.text:004F6DC5                 rep movsd
.text:004F6DC7                 call    eax             ; Does not modify esi.
.text:004F6DC9                 mov     eax, eax        ; The end of "KiSystemCallTrampoline".
.text:004F6DCB                 cmp     dword ptr [esi+0Ch], 1000h ; The code references the dword at (esi+0Ch).
.text:004F6DCB                                         ; The compiler could use esi like this if it knew
.text:004F6DCB                                         ;   something about the esi's value. esi is
.text:004F6DCB                                         ;   "Arguments" argument, we don't use it any
.text:004F6DCB                                         ;   place after in this function.
.text:004F6DCB                                         ; "Arguments" argument is travelled to esi like this:
.text:004F6DCB                                         ;   (the second argument of the function) edx ->
.text:004F6DCB                                         ;   [esi+14h] -> edi -> edx
.text:004F6DCB                                         ;   ->(just before the inline assembly block) esi.
.text:004F6DCB                                         ; Looks like the compiler meant to continue using esi
.text:004F6DCB                                         ;   as a register for the stack frame.
.text:004F6DD2                 jb      short loc_4F6DE4
.text:004F6DD4                 mov     ecx, ds:dword_59700C
.text:004F6DDA                 test    ecx, ecx
.text:004F6DDC                 jz      short loc_4F6DE4
.text:004F6DDE                 push    eax
.text:004F6DDF                 push    dword ptr [esi+0Ch]
.text:004F6DE2                 call    ecx ; dword_59700C
.text:004F6DE4
.text:004F6DE4 loc_4F6DE4:                             ; CODE XREF: sub_4F6BD0+202↑j
.text:004F6DE4                                         ; sub_4F6BD0+20C↑j
.text:004F6DE4                 mov     edx, [esi+4]
.text:004F6DE7                 test    byte ptr [ebx+6Ch], 1
.text:004F6DEB                 mov     [esi], eax
.text:004F6DED                 jz      short loc_4F6E55
.text:004F6DEF                 call    ds:KeGetCurrentIrql
.text:004F6DF5                 test    al, al
.text:004F6DF7                 jnz     short loc_4F6E67
.text:004F6DF9                 mov     eax, ds:0FFDFF124h
.text:004F6DFE                 movzx   ecx, byte ptr [eax+11Ch]
.text:004F6E05                 test    ecx, ecx
.text:004F6E07                 mov     edx, [esi+4]
.text:004F6E0A                 jnz     short loc_4F6E12
.text:004F6E0C                 cmp     dword ptr [eax+70h], 0
.text:004F6E10                 jz      short loc_4F6E55
.text:004F6E12
.text:004F6E12 loc_4F6E12:                             ; CODE XREF: sub_4F6BD0+23A↑j
.text:004F6E12                 push    0
.text:004F6E14                 push    dword ptr [eax+70h]
.text:004F6E17                 push    ecx
.text:004F6E18                 push    dword ptr [esi+8]
.text:004F6E1B                 push    1
.text:004F6E1D                 call    KeBugCheckEx
.text:004F6E22
.text:004F6E22 loc_4F6E22:                             ; CODE XREF: sub_4F6BD0+1AF↑j
.text:004F6E22                 mov     dword ptr [esi], 0C000001Ch
.text:004F6E28                 call    sub_403FC1
.text:004F6E2D                 mov     edx, [esi+4]    ; The problem was happening here for me
.text:004F6E30                 mov     ebx, [edx+110h] ; or here.
.text:004F6E36                 mov     ecx, [edx+118h]
.text:004F6E3C                 add     ecx, edi
.text:004F6E3E                 test    eax, eax
.text:004F6E40                 js      short loc_4F6E53
.text:004F6E42                 mov     eax, [esi+8]
.text:004F6E45                 mov     [esi+10h], ecx
.text:004F6E48                 cmp     eax, [ecx+8]
.text:004F6E4B                 jb      loc_4F6CF8
.text:004F6E51                 jmp     short loc_4F6E55
.text:004F6E53 ; ---------------------------------------------------------------------------
.text:004F6E53
.text:004F6E53 loc_4F6E53:                             ; CODE XREF: sub_4F6BD0+270↑j
.text:004F6E53                 mov     [esi], eax
.text:004F6E55
.text:004F6E55 loc_4F6E55:                             ; CODE XREF: sub_4F6BD0+1BE↑j
.text:004F6E55                                         ; sub_4F6BD0+21D↑j ...
.text:004F6E55                 mov     eax, [ebx+3Ch]
.text:004F6E58                 mov     [edx+110h], eax
.text:004F6E5E                 mov     ecx, ebx
.text:004F6E60                 mov     edx, [esi]
.text:004F6E62                 call    sub_4F3E90
.text:004F6E67
.text:004F6E67 loc_4F6E67:                             ; CODE XREF: sub_4F6BD0+227↑j
.text:004F6E67                 movzx   eax, al
.text:004F6E6A                 mov     byte ptr ds:0FFDFF024h, 0
.text:004F6E71                 cli
.text:004F6E72                 push    0
.text:004F6E74                 push    0
.text:004F6E76                 push    eax
.text:004F6E77                 push    dword ptr [esi+8]
.text:004F6E7A                 push    4Ah
.text:004F6E7C                 call    KeBugCheckEx
.text:004F6E81                 nop
.text:004F6E82                 nop
.text:004F6E83                 nop
.text:004F6E84                 nop
.text:004F6E85                 nop
.text:004F6E86                 nop
.text:004F6E87                 nop
.text:004F6E88                 nop
.text:004F6E89                 nop
.text:004F6E8A                 nop
.text:004F6E8B                 nop
.text:004F6E8C                 nop
.text:004F6E8D                 nop
.text:004F6E8E                 nop
.text:004F6E8F                 nop
.text:004F6E8F sub_4F6BD0      endp ; sp-analysis failed
```
</details>

<details>
<summary>Disassembly after the change</summary>

```
.text:004F6BD0 ; =============== S U B R O U T I N E =======================================
.text:004F6BD0
.text:004F6BD0 ; Attributes: bp-based frame fuzzy-sp
.text:004F6BD0
.text:004F6BD0 sub_4F6BD0      proc near               ; CODE XREF: sub_403D39+62↑p
.text:004F6BD0                                         ; sub_403D39+F6↑p ...
.text:004F6BD0
.text:004F6BD0 var_3C          = dword ptr -3Ch
.text:004F6BD0 var_38          = dword ptr -38h
.text:004F6BD0 var_34          = dword ptr -34h
.text:004F6BD0 var_30          = dword ptr -30h
.text:004F6BD0 var_28          = dword ptr -28h
.text:004F6BD0 var_24          = dword ptr -24h
.text:004F6BD0 var_20          = dword ptr -20h
.text:004F6BD0 Id              = dword ptr -1Ch
.text:004F6BD0 var_18          = dword ptr -18h
.text:004F6BD0 SystemCallNumber= dword ptr -14h
.text:004F6BD0 Arguments       = dword ptr -10h
.text:004F6BD0
.text:004F6BD0                 push    ebp
.text:004F6BD1                 mov     ebp, esp
.text:004F6BD3                 push    ebx
.text:004F6BD4                 push    edi
.text:004F6BD5                 push    esi
.text:004F6BD6                 and     esp, 0FFFFFFFCh
.text:004F6BD9                 sub     esp, 14h
.text:004F6BDC                 mov     [esp+20h+Arguments], edx
.text:004F6BE0                 mov     ebx, ecx
.text:004F6BE2                 mov     eax, [ecx+3Ch]  ; eax <- TrapFrame->Edx
.text:004F6BE5                 mov     esi, [ecx+44h]  ; esi <- TrapFrame->Eax
.text:004F6BE8                 mov     edx, ds:0FFDFF124h
.text:004F6BEE                 mov     [ecx+0Ch], eax
.text:004F6BF1                 mov     dword ptr [ecx+8], 0BADB0D00h
.text:004F6BF8                 mov     eax, [ecx+68h]
.text:004F6BFB                 mov     [ecx+4], eax
.text:004F6BFE                 mov     eax, [ecx+60h]
.text:004F6C01                 mov     [ecx], eax
.text:004F6C03                 mov     dword ptr [ecx+48h], 0FFFFFFFFh
.text:004F6C0A                 mov     eax, [edx+110h]
.text:004F6C10                 mov     [ecx+3Ch], eax
.text:004F6C13                 mov     dword ptr [ecx+64h], 0
.text:004F6C1A                 movsx   eax, byte ptr [edx+0D7h]
.text:004F6C21                 mov     [ecx+48h], eax
.text:004F6C24                 mov     eax, ds:0FFDFF000h
.text:004F6C29                 mov     [ecx+4Ch], eax
.text:004F6C2C                 mov     dword ptr ds:0FFDFF000h, 0FFFFFFFFh
.text:004F6C36                 mov     dword ptr [ecx+2Ch], 0
.text:004F6C3D                 test    byte ptr [ecx+6Ch], 1
.text:004F6C41                 jz      short loc_4F6CB3
.text:004F6C43                 mov     eax, ds:0FFDFF124h
.text:004F6C48                 cmp     byte ptr [eax+3], 0
.text:004F6C4C                 jz      short loc_4F6CB3
.text:004F6C4E                 mov     eax, large fs:20h
.text:004F6C54                 mov     ecx, dr0
.text:004F6C57                 mov     [ebx+18h], ecx
.text:004F6C5A                 mov     ecx, dr1
.text:004F6C5D                 mov     [ebx+1Ch], ecx
.text:004F6C60                 mov     ecx, dr2
.text:004F6C63                 mov     [ebx+20h], ecx
.text:004F6C66                 mov     ecx, dr3
.text:004F6C69                 mov     [ebx+24h], ecx
.text:004F6C6C                 mov     ecx, dr6
.text:004F6C6F                 mov     [ebx+28h], ecx
.text:004F6C72                 mov     ecx, dr7
.text:004F6C75                 mov     [ebx+2Ch], ecx
.text:004F6C78                 xor     ecx, ecx
.text:004F6C7A                 mov     dr7, ecx
.text:004F6C7D                 mov     ecx, [eax+2F8h]
.text:004F6C83                 mov     dr0, ecx
.text:004F6C86                 mov     ecx, [eax+2FCh]
.text:004F6C8C                 mov     dr1, ecx
.text:004F6C8F                 mov     ecx, [eax+300h]
.text:004F6C95                 mov     dr2, ecx
.text:004F6C98                 mov     ecx, [eax+304h]
.text:004F6C9E                 mov     dr3, ecx
.text:004F6CA1                 mov     ecx, [eax+308h]
.text:004F6CA7                 mov     dr6, ecx
.text:004F6CAA                 mov     eax, [eax+30Ch]
.text:004F6CB0                 mov     dr7, eax
.text:004F6CB3
.text:004F6CB3 loc_4F6CB3:                             ; CODE XREF: sub_4F6BD0+71↑j
.text:004F6CB3                                         ; sub_4F6BD0+7C↑j
.text:004F6CB3                 mov     [edx+110h], ebx
.text:004F6CB9                 mov     al, [ebx+6Ch]
.text:004F6CBC                 and     al, 1
.text:004F6CBE                 mov     [edx+0D7h], al
.text:004F6CC4                 sti
.text:004F6CC5                 mov     edi, esi
.text:004F6CC7                 shr     edi, 8
.text:004F6CCA                 and     edi, 10h
.text:004F6CCD                 mov     [esp+20h+SystemCallNumber], esi ; SystemCallNumber <- TrapFrame->Edx
.text:004F6CD1                 mov     eax, esi
.text:004F6CD3                 and     eax, 0FFFh      ; eax <- SystemCallNumber & SERVICE_NUMBER_MASK
.text:004F6CD8                 mov     esi, [edx+118h]
.text:004F6CDE                 mov     ecx, eax
.text:004F6CE0                 mov     [esp+20h+Id], eax ; Id = SystemCallNumber & SERVICE_NUMBER_MASK
.text:004F6CE4                 cmp     eax, [esi+edi+8]
.text:004F6CE8                 jnb     loc_4F6D80
.text:004F6CEE                 add     esi, edi
.text:004F6CF0
.text:004F6CF0 loc_4F6CF0:                             ; CODE XREF: sub_4F6BD0+1E5↓j
.text:004F6CF0                 test    edi, edi
.text:004F6CF2                 jz      short loc_4F6D0E
.text:004F6CF4                 mov     eax, large fs:18h
.text:004F6CFA                 cmp     dword ptr [eax+0F70h], 0
.text:004F6D01                 jz      short loc_4F6D0E
.text:004F6D03                 mov     eax, ds:off_58D848
.text:004F6D08                 mov     edi, edx
.text:004F6D0A                 call    dword ptr [eax]
.text:004F6D0C                 mov     edx, edi
.text:004F6D0E
.text:004F6D0E loc_4F6D0E:                             ; CODE XREF: sub_4F6BD0+122↑j
.text:004F6D0E                                         ; sub_4F6BD0+131↑j
.text:004F6D0E                 mov     eax, large fs:20h
.text:004F6D14                 inc     dword ptr [eax+598h]
.text:004F6D1A                 mov     eax, [esi+0Ch]
.text:004F6D1D                 mov     ecx, [esp+20h+Id]
.text:004F6D21                 movzx   edi, byte ptr [eax+ecx]
.text:004F6D25                 mov     eax, ds:off_58DBC4
.text:004F6D2A                 mov     ecx, [esp+20h+Arguments]
.text:004F6D2E                 cmp     [eax], ecx
.text:004F6D30                 jbe     loc_4F6DC0
.text:004F6D36                 test    byte ptr [ebx+6Ch], 1
.text:004F6D3A                 jnz     loc_4F6DC0
.text:004F6D40                 push    72Ch
.text:004F6D45                 push    offset aNtoskrnlKeI386_7 ; "ntoskrnl/ke/i386/traphdlr.c"
.text:004F6D4A                 push    offset aKisystemservic ; "KiSystemServiceHandler"
.text:004F6D4F                 push    offset aUnrecoverableE_0 ; "UNRECOVERABLE ERROR:  %s at %s:%d is UN"...
.text:004F6D54                 call    DbgPrint
.text:004F6D59                 add     esp, 10h
.text:004F6D5C                 push    offset unk_56DC42
.text:004F6D61                 call    DbgPrint
.text:004F6D66                 add     esp, 4
.text:004F6D69                 call    DbgBreakPoint
.text:004F6D6E                 push    0
.text:004F6D70                 push    0
.text:004F6D72                 push    0
.text:004F6D74                 push    0C0000002h
.text:004F6D79                 push    5Ah
.text:004F6D7B                 call    KeBugCheckEx
.text:004F6D80
.text:004F6D80 loc_4F6D80:                             ; CODE XREF: sub_4F6BD0+118↑j
.text:004F6D80                 mov     [esp+34h+var_34], 0C000001Ch
.text:004F6D87                 test    edi, edi
.text:004F6D89                 jz      loc_4F6E62
.text:004F6D8F                 mov     esi, edx
.text:004F6D91                 call    sub_403FC1
.text:004F6D96                 mov     edx, esi
.text:004F6D98                 mov     ebx, [esi+110h]
.text:004F6D9E                 mov     esi, [esi+118h]
.text:004F6DA4                 add     esi, edi
.text:004F6DA6                 test    eax, eax
.text:004F6DA8                 js      loc_4F6E5F
.text:004F6DAE                 mov     eax, [esp+34h+var_30]
.text:004F6DB2                 cmp     eax, [esi+8]
.text:004F6DB5                 jb      loc_4F6CF0
.text:004F6DBB                 jmp     loc_4F6E62
.text:004F6DC0 ; ---------------------------------------------------------------------------
.text:004F6DC0
.text:004F6DC0 loc_4F6DC0:                             ; CODE XREF: sub_4F6BD0+160↑j
.text:004F6DC0                                         ; sub_4F6BD0+16A↑j
.text:004F6DC0                 mov     [esp+20h+var_18], edx
.text:004F6DC4                 cmp     [esp+20h+SystemCallNumber], 1000h
.text:004F6DCC                 jb      short loc_4F6DE1
.text:004F6DCE                 mov     eax, ds:dword_597008
.text:004F6DD3                 test    eax, eax
.text:004F6DD5                 jz      short loc_4F6DE1
.text:004F6DD7                 push    [esp+20h+Arguments]
.text:004F6DDB                 push    [esp+24h+SystemCallNumber]
.text:004F6DDF                 call    eax ; dword_597008
.text:004F6DE1
.text:004F6DE1 loc_4F6DE1:                             ; CODE XREF: sub_4F6BD0+1FC↑j
.text:004F6DE1                                         ; sub_4F6BD0+205↑j
.text:004F6DE1                 mov     eax, [esi]
.text:004F6DE3                 mov     ecx, [esp+28h+var_24]
.text:004F6DE7                 mov     eax, [eax+ecx*4]
.text:004F6DEA                 mov     ecx, edi
.text:004F6DEC                 mov     edx, [esp+28h+var_18]
.text:004F6DF0                 sub     esp, ecx        ; The start of "KiSystemCallTrampoline".
.text:004F6DF2                 mov     edi, esp
.text:004F6DF4                 mov     esi, edx
.text:004F6DF6                 shr     ecx, 2
.text:004F6DF9                 rep movsd
.text:004F6DFB                 call    eax
.text:004F6DFD                 mov     edx, eax        ; The end of "KiSystemCallTrampoline".
.text:004F6DFF                 mov     eax, [esp+28h+Id] ; After the changes, the compiler references "Id" here.
.text:004F6E03                 cmp     eax, 1000h
.text:004F6E08                 jb      short loc_4F6E1D
.text:004F6E0A                 mov     ecx, ds:dword_59700C
.text:004F6E10                 test    ecx, ecx
.text:004F6E12                 jz      short loc_4F6E1D
.text:004F6E14                 push    edx
.text:004F6E15                 push    eax
.text:004F6E16                 call    ecx ; dword_59700C
.text:004F6E18                 mov     [esp+30h+var_30], eax
.text:004F6E1B                 jmp     short loc_4F6E20
.text:004F6E1D ; ---------------------------------------------------------------------------
.text:004F6E1D
.text:004F6E1D loc_4F6E1D:                             ; CODE XREF: sub_4F6BD0+238↑j
.text:004F6E1D                                         ; sub_4F6BD0+242↑j
.text:004F6E1D                 mov     [esp+28h+var_28], edx
.text:004F6E20
.text:004F6E20 loc_4F6E20:                             ; CODE XREF: sub_4F6BD0+24B↑j
.text:004F6E20                 mov     edx, [esp+28h+var_20]
.text:004F6E24                 test    byte ptr [ebx+6Ch], 1
.text:004F6E28                 jz      short loc_4F6E62
.text:004F6E2A                 call    ds:KeGetCurrentIrql
.text:004F6E30                 test    al, al
.text:004F6E32                 jnz     short loc_4F6E75
.text:004F6E34                 mov     eax, ds:0FFDFF124h
.text:004F6E39                 movzx   ecx, byte ptr [eax+11Ch]
.text:004F6E40                 test    ecx, ecx
.text:004F6E42                 mov     edx, [esp+28h+var_20]
.text:004F6E46                 jnz     short loc_4F6E4E
.text:004F6E48                 cmp     dword ptr [eax+70h], 0
.text:004F6E4C                 jz      short loc_4F6E62
.text:004F6E4E
.text:004F6E4E loc_4F6E4E:                             ; CODE XREF: sub_4F6BD0+276↑j
.text:004F6E4E                 push    0
.text:004F6E50                 push    dword ptr [eax+70h]
.text:004F6E53                 push    ecx
.text:004F6E54                 push    [esp+34h+var_24]
.text:004F6E58                 push    1
.text:004F6E5A                 call    KeBugCheckEx
.text:004F6E5F
.text:004F6E5F loc_4F6E5F:                             ; CODE XREF: sub_4F6BD0+1D8↑j
.text:004F6E5F                 mov     [esp+3Ch+var_3C], eax
.text:004F6E62
.text:004F6E62 loc_4F6E62:                             ; CODE XREF: sub_4F6BD0+1B9↑j
.text:004F6E62                                         ; sub_4F6BD0+1EB↑j ...
.text:004F6E62                 mov     eax, [ebx+3Ch]
.text:004F6E65                 mov     [edx+110h], eax
.text:004F6E6B                 mov     ecx, ebx
.text:004F6E6D                 mov     edx, [esp+3Ch+var_3C]
.text:004F6E70                 call    sub_4F3E90
.text:004F6E75
.text:004F6E75 loc_4F6E75:                             ; CODE XREF: sub_4F6BD0+262↑j
.text:004F6E75                 movzx   eax, al
.text:004F6E78                 mov     byte ptr ds:0FFDFF024h, 0
.text:004F6E7F                 cli
.text:004F6E80                 push    0
.text:004F6E82                 push    0
.text:004F6E84                 push    eax
.text:004F6E85                 push    [esp+48h+var_38]
.text:004F6E89                 push    4Ah
.text:004F6E8B                 call    KeBugCheckEx
.text:004F6E8B sub_4F6BD0      endp ; sp-analysis failed
```
</details>

Reply:
It doesn't looks like an assembly output for `KiSystemCallTrampoline`

Reply:
Those are disassemblies of `KiSystemServiceHandler`. `KiSystemCallTrampoline` is always inlined and the only usage (according to the search query `KiSystemCallTrampoline` ([link](https://git.reactos.org/?p=reactos.git&a=search&h=206156fad57dc951cde3e100e584add0d1e8e4ec&st=grep&s=KiSystemCallTrampoline))) is in `KiSystemServiceHandler`. I failed to find `KiSystemCallTrampoline` in a disassembly of `ntoskrnl.exe`.

It might be that the commit message should be written in a different way (it doesn't tell that the problem was in the function that is using `KiSystemCallTrampoline`). Tell me if it so, I should rewrite it then.

Reply:
See PR #3588